File: docker/hik-sdk/Samples/64/C++/LineScanCamera/LineScanSoftwareTrigger/LineScanSoftwareTrigger.cpp
/*
* 这个示例演示线阵相机如何通过软触发命令获取图像
* This simple shows how a line-scan camera acquires images through software trigger commands.
*
*/
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <pthread.h>
#include <string.h>
#include "MvCameraControl.h"

bool g_bExit = false;
char g_chTriggerCmd[128];

// ch: 检查属性节点的访问模式 | en: Check the access mode of the feature node
bool CheckFeatureNodeAccess(void* hHandle, const char* pNodeName);
// ch：打印相机信息 | en: Print the information of cameras
bool PrintDeviceInfo(MV_CC_DEVICE_INFO* pstMVDevInfo);
// ch: 取图回调函数 | en: Image callback function
void __stdcall ImageCallbackEx2(MV_FRAME_OUT* pstFrame, void *pUser, bool bAutoFree);
// ch: 软触发命令线程 | en: software trigger command thread
static  void* SoftwareTriggerCommandThread(void* pUser);

// 等待用户输入enter键来结束取流或结束程序
// wait for user to input enter to stop grabbing or end the sample program
void PressEnterToExit(void)
{
    int c;
    while ( (c = getchar()) != '\n' && c != EOF );
    fprintf( stderr, "\nPress enter to exit.\n");
    while( getchar() != '\n');
}

int main()
{
    int nRet = MV_OK;
    void* handle = NULL;

    do 
    {
        // ch:初始化SDK | en:Initialize SDK
        nRet = MV_CC_Initialize();
        if (MV_OK != nRet)
        {
            printf("Initialize SDK fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:枚举设备 | en:Enum device
        MV_CC_DEVICE_INFO_LIST stDeviceList;
        memset(&stDeviceList, 0, sizeof(MV_CC_DEVICE_INFO_LIST));
        nRet = MV_CC_EnumDevices(MV_GIGE_DEVICE | MV_USB_DEVICE | MV_GENTL_CAMERALINK_DEVICE | MV_GENTL_CXP_DEVICE | MV_GENTL_XOF_DEVICE, &stDeviceList);
        if (MV_OK != nRet)
        {
            printf("Enum Devices fail! nRet [0x%x]\n", nRet);
            break;
        }

        if (stDeviceList.nDeviceNum > 0)
        {
            for (unsigned int i = 0; i < stDeviceList.nDeviceNum; i++)
            {
                printf("[device %d]:\n", i);
                MV_CC_DEVICE_INFO* pDeviceInfo = stDeviceList.pDeviceInfo[i];
                if (NULL == pDeviceInfo)
                {
                    break;
                } 
                PrintDeviceInfo(pDeviceInfo);
            }  
        } 
        else
        {
            printf("Find No Devices!\n");
            break;
        }

        printf("Please Input camera index(0-%d):", stDeviceList.nDeviceNum-1);
        unsigned int nIndex = 0;
        scanf("%d", &nIndex);

        if (nIndex >= stDeviceList.nDeviceNum)
        {
            printf("Input error!\n");
            break;
        }

        // ch:选择设备并创建句柄 | en:Select device and create handle
        nRet = MV_CC_CreateHandle(&handle, stDeviceList.pDeviceInfo[nIndex]);
        if (MV_OK != nRet)
        {
            printf("Create Handle fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:打开设备 | en:Open device
        nRet = MV_CC_OpenDevice(handle);
        if (MV_OK != nRet)
        {
            printf("Open Device fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:探测网络最佳包大小(只对GigE相机有效) | en:Detection network optimal package size(It only works for the GigE camera)
        if (stDeviceList.pDeviceInfo[nIndex]->nTLayerType == MV_GIGE_DEVICE)
        {
            int nPacketSize = MV_CC_GetOptimalPacketSize(handle);
            if (nPacketSize > 0)
            {
                nRet = MV_CC_SetIntValueEx(handle,"GevSCPSPacketSize",nPacketSize);
                if(nRet != MV_OK)
                {
                    printf("Warning: Set Packet Size fail nRet [0x%x]!\n", nRet);
                }
            }
            else
            {
                printf("Warning: Get Packet Size fail nRet [0x%x]!\n", nPacketSize);
            }
        }

        nRet = MV_CC_SetEnumValueByString(handle, "ScanMode", "FrameScan");
        if (MV_OK == nRet)
        {
            printf("Set Frame Scan Mode \n");
        }

        // ch: 判断FrameTriggerControl是否可读 | en: Check if FrameTriggerControl is readable
        if (CheckFeatureNodeAccess(handle, "FrameTriggerControl"))
        {
            // ch:设置触发模式为on | en:Set trigger mode as on
            nRet = MV_CC_SetBoolValue(handle, "FrameTriggerMode", true);
            if (MV_OK != nRet)
            {
                printf("Set Frame Trigger Mode fail! nRet [0x%x]\n", nRet);
                break;
            }

            // ch:设置触发源为软触发 | en:Set trigger source as Software
            nRet = MV_CC_SetEnumValueByString(handle, "FrameTriggerSource", "Software");
            if (MV_OK != nRet)
            {
                printf("Set Frame Trigger source fail! nRet [0x%x]\n", nRet);
                break;
            }

            strcpy(g_chTriggerCmd, "FrameTriggerSoftware");
        }
        else
        {
            // ch:设置触发选项为FrameBurstStart | en:Set trigger selector as FrameBurstStart
            nRet = MV_CC_SetEnumValue(handle, "TriggerSelector", 6);
            if (MV_OK != nRet)
            {
                printf("Set Trigger Selector fail! nRet [0x%x]\n", nRet);
                break;
            }

            // ch:设置触发模式为on | en:Set trigger mode as on
            nRet = MV_CC_SetEnumValue(handle, "TriggerMode", 1);
            if (MV_OK != nRet)
            {
                printf("Set Trigger Mode fail! nRet [0x%x]\n", nRet);
                break;
            }

            // ch:设置触发源为Software | en:Set trigger source as Software
            nRet = MV_CC_SetEnumValueByString(handle, "TriggerSource", "Software");
            if (MV_OK != nRet)
            {
                printf("Set Trigger source fail! nRet [0x%x]\n", nRet);
                break;
            }

            strcpy(g_chTriggerCmd, "TriggerSoftware");
        }

        // ch:注册抓图回调 | en:Register image callback
        nRet = MV_CC_RegisterImageCallBackEx2(handle, ImageCallbackEx2, handle, true);
        if (MV_OK != nRet)
        {
            printf("Register Image CallBack fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:开始取流 | en:Start grab image
        nRet = MV_CC_StartGrabbing(handle);
        if (MV_OK != nRet)
        {
            printf("Start Grabbing fail! nRet [0x%x]\n", nRet);
            break;
        }
        
        pthread_t nThreadID;
        nRet = pthread_create(&nThreadID, NULL ,SoftwareTriggerCommandThread , handle);
        if (nRet != 0)
        {
            printf("thread create failed.ret = %d\n",nRet);
            break;
        }

        PressEnterToExit();

        g_bExit = true;
        sleep(1);

        // ch:停止取流 | en:Stop grab image
        nRet = MV_CC_StopGrabbing(handle);
        if (MV_OK != nRet)
        {
            printf("Stop Grabbing fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:注销抓图回调 | en:Unregister image callback
        nRet = MV_CC_RegisterImageCallBackEx2(handle, NULL, NULL, true);
        if (MV_OK != nRet)
        {
            printf("Unregister Image CallBack fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:关闭设备 | Close device
        nRet = MV_CC_CloseDevice(handle);
        if (MV_OK != nRet)
        {
            printf("ClosDevice fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:销毁句柄 | Destroy handle
        nRet = MV_CC_DestroyHandle(handle);
        if (MV_OK != nRet)
        {
            printf("Destroy Handle fail! nRet [0x%x]\n", nRet);
            break;
        }
        handle = NULL;
    } while (0);
 
    if (handle != NULL)
    {
        MV_CC_DestroyHandle(handle);
        handle = NULL;
    }
    

    // ch:反初始化SDK | en:Finalize SDK
    MV_CC_Finalize();


    return 0;
}

bool PrintDeviceInfo(MV_CC_DEVICE_INFO* pstMVDevInfo)
{
    if (NULL == pstMVDevInfo)
    {
        printf("The Pointer of pstMVDevInfo is NULL!\n");
        return false;
    }
    if (pstMVDevInfo->nTLayerType == MV_GIGE_DEVICE)
    {
        int nIp1 = ((pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0xff000000) >> 24);
        int nIp2 = ((pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0x00ff0000) >> 16);
        int nIp3 = ((pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0x0000ff00) >> 8);
        int nIp4 = (pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0x000000ff);

        // ch:打印当前相机ip和用户自定义名字 | en:print current ip and user defined name
        printf("CurrentIp: %d.%d.%d.%d\n", nIp1, nIp2, nIp3, nIp4);
        printf("UserDefinedName: %s\n\n", pstMVDevInfo->SpecialInfo.stGigEInfo.chUserDefinedName);
    }
    else if (pstMVDevInfo->nTLayerType == MV_USB_DEVICE)
    {
        printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stUsb3VInfo.chUserDefinedName);
        printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stUsb3VInfo.chSerialNumber);
        printf("Device Number: %d\n\n", pstMVDevInfo->SpecialInfo.stUsb3VInfo.nDeviceNumber);
    }
    else if (pstMVDevInfo->nTLayerType == MV_GENTL_GIGE_DEVICE)
    {
        printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stGigEInfo.chUserDefinedName);
        printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stGigEInfo.chSerialNumber);
        printf("Model Name: %s\n\n", pstMVDevInfo->SpecialInfo.stGigEInfo.chModelName);
    }
    else if (pstMVDevInfo->nTLayerType == MV_GENTL_CAMERALINK_DEVICE)
    {
        printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stCMLInfo.chUserDefinedName);
        printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stCMLInfo.chSerialNumber);
        printf("Model Name: %s\n\n", pstMVDevInfo->SpecialInfo.stCMLInfo.chModelName);
    }
    else if (pstMVDevInfo->nTLayerType == MV_GENTL_CXP_DEVICE)
    {
        printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stCXPInfo.chUserDefinedName);
        printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stCXPInfo.chSerialNumber);
        printf("Model Name: %s\n\n", pstMVDevInfo->SpecialInfo.stCXPInfo.chModelName);
    }
    else if (pstMVDevInfo->nTLayerType == MV_GENTL_XOF_DEVICE)
    {
        printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stXoFInfo.chUserDefinedName);
        printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stXoFInfo.chSerialNumber);
        printf("Model Name: %s\n\n", pstMVDevInfo->SpecialInfo.stXoFInfo.chModelName);
    }
    else
    {
        printf("Not support.\n");
    }

    return true;
}


void __stdcall ImageCallbackEx2(MV_FRAME_OUT* pstFrame, void *pUser, bool bAutoFree)
{
    if (pstFrame)
    {
        printf("Get One Frame: Width[%d], Height[%d], nFrameNum[%d]\n", 
            pstFrame->stFrameInfo.nExtendWidth, pstFrame->stFrameInfo.nExtendHeight, pstFrame->stFrameInfo.nFrameNum);

        if (false == bAutoFree &&
            NULL != pUser) //非自动释放模式，需要手动释放资源
        {
            MV_CC_FreeImageBuffer(pUser, pstFrame);
        }
    }
}


//ch: 检查属性节点的访问模式 | en: Check the access mode of the feature node
bool CheckFeatureNodeAccess(void* hHandle, const char* pNodeName)
{
    if (NULL == pNodeName|| NULL == hHandle)
    {
        return false;
    }

    MV_XML_AccessMode nMode;
    unsigned int nRet = MV_XML_GetNodeAccessMode(hHandle, pNodeName, &nMode);
    if (nRet != MV_OK)
    {
        return false;
    }

    if (nMode == AM_WO || nMode == AM_RO || nMode == AM_RW)
    {
        return true;
    }
    else
    {
        return false;
    }
}

// ch: 发送软触发命令 | en: send software trigger command
static void* SoftwareTriggerCommandThread(void* pUser)
{
    int nRet = MV_OK;

    if (pUser == NULL)
    {
        return 0;
    }

    while (true)
    {
        nRet = MV_CC_SetCommandValue(pUser, g_chTriggerCmd);
        if (nRet != MV_OK)
        {
            printf("Set software trigger command fail! nRet [0x%x]\n", nRet);
        }
        
        sleep(1);
        
        if (g_bExit)
        {
            break;
        }
    }

    return 0;
}

File: docker/hik-sdk/Samples/64/C#/MvCameraControlNet_V2/LineScanCamera/LineScanSoftwareTrigger/LineScanSoftwareTrigger.cs
/*
 * 这个示例演示线阵相机如何通过软触发命令获取图像
 * This simple shows how a line-scan camera acquires images through software trigger commands.
 */

using MvCameraControl;
using System;
using System.Collections.Generic;
using System.Threading;

namespace LineScanSoftwareTrigger
{
    class LineScanSoftwareTrigger
    {
        IDevice device = null;
        string triggerCmd="";

        const DeviceTLayerType devLayerType = DeviceTLayerType.MvGigEDevice | DeviceTLayerType.MvUsbDevice | DeviceTLayerType.MvGenTLCameraLinkDevice
            | DeviceTLayerType.MvGenTLCXPDevice | DeviceTLayerType.MvGenTLXoFDevice;

        volatile bool _cmdThreadExit = false;
        void SoftwareTriggerCommandThread(object obj)
        {
            IParameters parameters = (IParameters)obj;

            while (!_cmdThreadExit)
            {
                //ch：发送软触发命令 | en: set software trigger command
                int ret = parameters.SetCommandValue(triggerCmd);
                if (ret != MvError.MV_OK)
                {
                    Console.WriteLine("Set trigger command fail:{0:x8}", ret);
                    continue;
                }

                Thread.Sleep(1000);
            }
        }

        void FrameGrabedEventHandler(object sender, FrameGrabbedEventArgs e)
        {
            Console.WriteLine("Get one frame: Width[{0}] , Height[{1}] , FrameNum[{2}]", e.FrameOut.Image.Width, e.FrameOut.Image.Height, e.FrameOut.FrameNum);
        }

        //ch：检查节点是否可访问 | en: check if node is accessible
        bool CheckFeatureNodeAccess(string nodeName)
        {
            XmlAccessMode nMode;
            int ret = device.Parameters.GetNodeAccessMode(nodeName, out nMode);
            if (ret != MvError.MV_OK)
            {
                return false;
            }
            if(nMode == XmlAccessMode.RO || nMode == XmlAccessMode.RW || nMode == XmlAccessMode.WO)
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        public void Run()
        {
            try
            {
                List<IDeviceInfo> devInfoList;

                // ch:枚举设备 | en:Enum device
                int ret = DeviceEnumerator.EnumDevices(devLayerType, out devInfoList);
                if (ret != MvError.MV_OK)
                {
                    Console.WriteLine("Enum device failed:{0:x8}", ret);
                    return;
                }

                Console.WriteLine("Enum device count : {0}", devInfoList.Count);

                if (0 == devInfoList.Count)
                {
                    return;
                }

                // ch:打印设备信息 en:Print device info
                int devIndex = 0;
                foreach (var devInfo in devInfoList)
                {
                    Console.WriteLine("[Device {0}]:", devIndex);
                    if (devInfo.TLayerType == DeviceTLayerType.MvGigEDevice || devInfo.TLayerType == DeviceTLayerType.MvVirGigEDevice || devInfo.TLayerType == DeviceTLayerType.MvGenTLGigEDevice)
                    {
                        IGigEDeviceInfo gigeDevInfo = devInfo as IGigEDeviceInfo;
                        uint nIp1 = ((gigeDevInfo.CurrentIp & 0xff000000) >> 24);
                        uint nIp2 = ((gigeDevInfo.CurrentIp & 0x00ff0000) >> 16);
                        uint nIp3 = ((gigeDevInfo.CurrentIp & 0x0000ff00) >> 8);
                        uint nIp4 = (gigeDevInfo.CurrentIp & 0x000000ff);
                        Console.WriteLine("DevIP: {0}.{1}.{2}.{3}", nIp1, nIp2, nIp3, nIp4);
                    }

                    Console.WriteLine("ModelName:" + devInfo.ModelName);
                    Console.WriteLine("SerialNumber:" + devInfo.SerialNumber);
                    Console.WriteLine();
                    devIndex++;
                }

                Console.Write("Please input index(0-{0:d}):", devInfoList.Count - 1);

                // ch:选择设备 | en:Select a device
                devIndex = Convert.ToInt32(Console.ReadLine());

                if (devIndex > devInfoList.Count - 1 || devIndex < 0)
                {
                    Console.Write("Input Error!\n");
                    return;
                }

                // ch:创建设备 | en:Create device
                device = DeviceFactory.CreateDevice(devInfoList[devIndex]);

                // ch:打开设备 | en:Open device
                ret = device.Open();
                if (ret != MvError.MV_OK)
                {
                    Console.WriteLine("Open device failed:{0:x8}", ret);
                    return;
                }
                Console.WriteLine("Open device success");

                // ch:探测网络最佳包大小(只对GigE相机有效) | en:Detection network optimal package size(It only works for the GigE camera)
                if (device is IGigEDevice)
                {
                    int packetSize;
                    ret = (device as IGigEDevice).GetOptimalPacketSize(out packetSize);
                    if (packetSize > 0)
                    {
                        ret = device.Parameters.SetIntValue("GevSCPSPacketSize", packetSize);
                        if (ret != MvError.MV_OK)
                        {
                            Console.WriteLine("Warning: Set Packet Size failed {0:x8}", ret);
                        }
                        else
                        {
                            Console.WriteLine("Set PacketSize to {0}", packetSize);
                        }
                    }
                    else
                    {
                        Console.WriteLine("Warning: Get Packet Size failed {0:x8}", ret);
                    }
                }

                ret = device.Parameters.SetEnumValueByString("ScanMode", "FrameScan");
                if (ret == MvError.MV_OK)
                {
                    Console.WriteLine("Set Frame Scan Mode ");
                }

                // ch: 检查FrameTriggerControl是否可读 | en: Check if FrameTriggerControl is readable
                if (CheckFeatureNodeAccess("FrameTriggerControl"))
                {
                    // ch:设置触发模式为on | en:Set trigger mode as On
                    ret = device.Parameters.SetBoolValue("FrameTriggerMode", true);
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set Frame Trigger Mode fail! nRet {0}\n", ret.ToString("X"));
                        return;
                    }
                    Console.WriteLine("Set Frame Trigger Mode to On");

                    // ch:设置触发源为Line0 | en:Set trigger source as software
                    ret = device.Parameters.SetEnumValueByString("FrameTriggerSource", "Software");
                    if (ret != MvError.MV_OK)
                    {
                         Console.WriteLine("Set Frame Trigger Source fail! nRet {0}\n", ret.ToString("X"));
                         return;
                    }
                    Console.WriteLine("Set Frame Trigger Source to software");

                    triggerCmd = "FrameTriggerSoftware";
                }
                else
                {
                    // ch:设置触发选项为FrameBurstStart | en:Set trigger selector as FrameBurstStart
                    ret = device.Parameters.SetEnumValue("TriggerSelector", 6);
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set Trigger Selector fail! nRet {0}\n", ret.ToString("X"));
                        return;
                    }
                    Console.WriteLine("Set Trigger Selector to FrameBurstStart");

                    // ch:设置触发模式为on | en:Set trigger mode as On
                    ret = device.Parameters.SetEnumValue("TriggerMode", 1);
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set Trigger Mode fail! nRet {0}\n", ret.ToString("X"));
                        return;
                    }
                    Console.WriteLine("Set Trigger Mode to On");

                    // ch:设置触发源为Line0 | en:Set trigger source as Software
                    ret = device.Parameters.SetEnumValueByString("TriggerSource", "Software");
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set Trigger Source fail! nRet {0}\n", ret.ToString("X"));
                        return;
                    }
                    Console.WriteLine("Set Trigger Source to software");

                    triggerCmd = "TriggerSoftware";
                }

                //ch: 设置合适的缓存节点数量 | en: Setting the appropriate number of image nodes
                device.StreamGrabber.SetImageNodeNum(5);

                // ch:注册回调函数 | en:Register image callback
                device.StreamGrabber.FrameGrabedEvent += FrameGrabedEventHandler;
                // ch:开启抓图 || en: start grab image
                ret = device.StreamGrabber.StartGrabbing();
                if (ret != MvError.MV_OK)
                {
                    Console.WriteLine("Start grabbing failed:{0:x8}", ret);
                    return;
                }

                // ch:开启执行触发命令线程 | en: Start set trigger command thread
                Thread CommandThread = new Thread(SoftwareTriggerCommandThread);
                CommandThread.Start(device.Parameters);

                Console.WriteLine("Press enter to Stop Grabbing");
                Console.ReadLine();

                //ch: 通知线程退出 | en: Notify the thread to exit
                _cmdThreadExit = true;
                CommandThread.Join();


                // ch:停止抓图 | en:Stop grabbing
                ret = device.StreamGrabber.StopGrabbing();
                if (ret != MvError.MV_OK)
                {
                    Console.WriteLine("Stop grabbing failed:{0:x8}", ret);
                    return;
                }

                // ch:关闭设备 | en:Close device
                ret = device.Close();
                if (ret != MvError.MV_OK)
                {
                    Console.WriteLine("Close device failed:{0:x8}", ret);
                    return;
                }

                // ch:销毁设备 | en:Destroy device
                device.Dispose();
                device = null;
            }
            catch (Exception e)
            {
                Console.Write("Exception: " + e.Message);
            }
            finally
            {
                // ch:销毁设备 | en:Destroy device
                if (device != null)
                {
                    device.Dispose();
                    device = null;
                }
            }
        }

        static void Main(string[] args)
        {
            // ch: 初始化 SDK | en: Initialize SDK
            SDKSystem.Initialize();

            LineScanSoftwareTrigger program = new LineScanSoftwareTrigger();
            program.Run();

            Console.WriteLine("Press enter to exit");
            Console.ReadKey();

            // ch: 反初始化SDK | en: Finalize SDK
            SDKSystem.Finalize();
        }
    }
}
